luigi/contrib/jupyter_notebook.py

```diff
@@ -0,0 +1,310 @@
+                   'Please install nbconvert and nbformat.')
+
+
+def _get_file_name_from_path(input_path):
+    """
+    A simple utility to extract the name of a file without the file extension
+    from a given path.
+
+    This function extracts the file name without the file extension from a given
+    path.
+    For example, if `path=~/some_dir/some_file.ext`, then
+    `get_filename_from_path(path)` will return `some_file`.
+
+    :param input_path: a path to a file
+
+    :returns: a string containing the name of the file without the file
+        extension
+    """
+    base_name = os.path.basename(os.path.normpath(input_path))
+    file_name = base_name.split('.')[0]
+    return file_name
+
+
+def _get_values(obj):
+    """
+    A simple utility to extract the values of a dictionary, list, or other
+    iterable and recombine them into a list.
+
+    :param obj: a dictionary, list, or other iterable
+
+    :returns: a list with the values of obj
+    """
+    if isinstance(obj, dict):
+        out = obj.values()
+    else:
+        out = [val for val in obj]
+    return out
+
+
+class JupyterNotebookTask(luigi.Task):
+    """
+    This is a template task to execute Jupyter notebooks as Luigi tasks.
+    This task has the following key parameters:
+
+    :param nb_path: the full path to the Jupyter notebook (**required**).
+
+    :param kernel_name: the name of the kernel to be used in the notebook
+        execution (**required**).
+
+    :param timeout: maximum time (in seconds) allocated to run each cell.
+        If -1 (the default), no timeout limit is imposed.
+
+    :param pars: a dictionary of user-defined parameters to be passed to the
+        Jupyter notebook.
+
+    :param json_action: if `delete` (default), the temporary JSON file with
+        **pars** is deleted at the end of the task execution; if `keep`, the
+        temporary JSON file is kept (useful for debugging purposes).
+    """
+    nb_path = luigi.Parameter(
+        default=None
+    )
+
+    kernel_name = luigi.Parameter(
+        default=None
+    )
+
+    timeout = luigi.IntParameter(
+        default=-1
+    )
+
+    json_action = luigi.Parameter(
+        default='delete'
+    )
+
+    pars = {}
+
+    def _form_input(self):
+
+        if isinstance(self.input(), dict):
+            out = {
+                tag: list(map(
+                    lambda x: x.path, _get_values(self.input().get(tag)))
+                ) for tag in self.input().keys()
+            }
+        else:
+            out = [
+                list(map(lambda x: x.path, _get_values(req)))
+                for req in self.input()
+            ]
+
+        return out
+
+    def _form_output(self):
+
+        if isinstance(self.output(), dict):
+            out = {
+                tag: req.path for tag, req in self.output().items()
+            }
+        else:
+            out = [req.path for req in self.output()]
```

Comment:
I forgot to mention (and based on our conversation you noticed and fixed): inputs and outputs may not be iterable.

Reply:
Yes, this was definitely a bug. It is fixed now by getting values safely by leveraging the `flatten` utility in `luigi.task` and treating the `dict` case separately from the `list`/iterable case.
Thanks for pointing this out!